### ISSUE TITLE
Discussion Thread Milestone B(1) - Send Message to Field Technician

### ISSUE THREAD

**Issue by bsaiken** [FWAJL/FieldWorkAssistantMVC#803]
Entering text in the left hand column will activate the “Send” button in order to send the message.  

All messages shall be stored in the database to build a discussion thread history. You will need to request a new table (e.g. “discussion”) to store the discussion thread history.  

The table shall include the columns:
discussion_id
task_id. 

A another table (“discussion_content”) will hold the messages for a discussion so the columns will be:
discussion_content_id, 
discussion_id, 
discussion_content_category_type (e.g. technician_id) 
discussion_content_category_value (id value of the discussion_content_category_type ) 
discussion_content_value (the actual message sent at a given time).  

The tables structure may be modified as needed.

The message shall be sent to the FT.  The method for delivering the message is at your (the developer) discretion.  This can be done via email, txt, integrated into the Field Technician interface, or some other method.  Initially, I prefer the most readily available, cost effective method - we may add more messaging options to the site in the future.  The FT shall have some simple (single click) method to respond.  

Discussions with Field Technicians can be text only.

**Comment by bsaiken** [comment]
@nadirhamid @FWAJL 
I got your email message.  
If you want to send an email as a notification to the recipient, use their stored email address.  
Under your login credentials (thread:thread), I inserted the Field Technician "Test Technician - messages"
That's my email address stored in `technician.technician_email`, so that would be the one to use.
Please let us know if you have questions.

**Comment by nadirhamid** [comment]
@bsaiken @FWAJL 

So the milestone was put on hold for half a day given some issues, I will
have it ready in the next half day

-- Speaking on the repo any updates there not functional as they require
the later javascript "discussion_manager.js" and "discussion_manager.css"

I will update soon

**Comment by FWAJL** [comment]
@nadirhamid @bsaiken 

A suggestion: if you do some of the work and commit it for us to review it to see if you are understanding the requirements and the coding structure of the application, it'd be beneficial to all of us. Then we can correct some or just give you tips as the application is rather large and has a lot of built-in functions.

So do small commits of code that works even if the code for the milestone is not all finished. It will avoid back and forth updates.

One of the team member did this and it worked well for everyone. Thanks for considering it.

**Comment by nadirhamid** [comment]
Status has been set to review @bsaiken @FWAJL

**Comment by FWAJL** [comment]
@bsaiken @nadirhamid 

As per your comment in #880, any database updates need to be requested. So I will make the changes you request as I think they are useful.

After that, we have a tool that generate the Dao classes automatically. 

We will comment more soon after looking at your commits.

**Comment by FWAJL** [comment]
@bsaiken @nadirhamid 

A general request, as follow up of our last comment in #880: issues are to be created once we decide that there is a need for it. We write the SOW milestones which can be one or several issues depending on the milestone. Then if one of these issues needs more work out of scope, we will consider adding the work to the issue (the case here for the database updates) or create new issue(s).
So you should always comment in the issue you are finding break point and we will tell what to do next.

Thank you for working with us in that workflow. It'll make us successful as we've been doing it for a few months now with 2 people. We can do it with you too.

**Comment by bsaiken** [comment]
@nadirhamid @FWAJL 
Nadir,
If you have committed for me to review, You need to change the status to `status: function review`
Remove the `status: in dev` (this means you are actively working on it.
You must also set the assignee to `bsaiken`, otherwise, I am not aware you are ready for the code to be reviewed.
Also, as I mentioned to you before, you cannot modify the `appsettings.xml` file, or the site will not run on my local computer for testing, and we would need to change the file for deployment on the server.

**Comment by bsaiken** [comment]
@nadirhamid @FWAJL 
I have attempted to test this code, and I cannot get it to work on my local machine.
Please test the code yourself before making a commit to ensure that it is working correctly.

**Comment by bsaiken** [comment]
@nadirhamid @FWAJL 
I was able to get the site to function and do some preliminary testing.  When I choose who to communicate with, I get the popup confirmation box below:

![screen shot 2015-04-16 at 11 02 21 pm](https://cloud.githubusercontent.com/assets/8228624/7195876/2901cea4-e48d-11e4-81e0-2d85f04a6e09.png)

Did you add this?  I do not know where it came from.  The popup confirmation box should not be there.  This was not a requested behavior.

**Comment by FWAJL** [comment]
@nadirhamid @bsaiken 

Nadir,

My first comments and requests about your code:

1. I see many files being created anad having no use or purpose... ex: `FieldWorkAssistantMVC/Applications/PMTool/Helpers/:w` or `FieldWorkAssistantMVC/Applications/PMTool/Helpers/Contr*`. Can you remove them please?
2. What is the use of `FieldWorkAssistantMVC/Applications/PMTool/Models/Dal/2`?
3. `FieldWorkAssistantMVC/Applications/PMTool/Models/Dal/DiscussionDal.php`=> Please create a separate file for each class found in this file.
3. At [line 160 in this file](https://github.com/FWAJL/FieldWorkAssistantMVC/blob/feature_thread/Applications/PMTool/Controllers/ActiveTaskController.php#L160),you coded `isset($_SESSION[ \Library\Enums\SessionKeys::CurrentDiscussion])` => do not use $_SESSION variable directly. You can use `$this->user()->get(\Library\Enums\SessionKeys::CurrentDiscussion)` instead.
4. Please create a DiscussionHelper where you will put the methods to use and reuse for all the Discussion work. We do that so the controller don't have all the logic and also to allow other controllers and helpers to call the DiscussionHelper method if need. Always pass on the instance of controller object instance to the helper. It is used to access the user()->get(key) property that works with and on the Session.  
5. I have a bug. Please fix it.
![screen shot 2015-04-17 at 07 20 04](https://cloud.githubusercontent.com/assets/7524388/7196444/35dbbf14-e4d2-11e4-842a-61bbe61d5d53.png)

Also, if you build a AJAX POST request (which you will certainly do soon), please be aware that the data sent in accessible via `$this->dataPost()` in each controller method that is called via AJAX.

All the GET values are accessible via $rq->getData(key) also in each controller method.

Thank you.

**Comment by FWAJL** [comment]
@nadirhamid @bsaiken 

I need to review the database updates to do. Bare with me a little and I'll submit the updates. Thank you.

**Comment by FWAJL** [comment]
@nadirhamid @bsaiken 

>@nadirhamid
>
>I will be adding the following fields to discussion, and discussion_content
>
>discussion:
>discussion_time:
>a time when the discussion was started
>
>discussion_content:
>discussion_content_is_receiver
>Whether a pm sent the message or it was a response
>discussion_content_time
>A time for the reply, or initial post

I have added a new column `discussion_start_time` of TIMESTAMP type and it is set with the current timestamp at insertion.

I have added a new column `discussion_content_time` of TIMESTAMP type and it is set with the current timestamp at insertion.

However, I do not see the point of `discussion_content_is_receiver`? Can you explain?
In the table structure, the `discussion_content_category_type` value says if it is a pm or technician or any other person.

I have generated the updates Dao classes. Please integrate them in your code and let me know if you have question.

Thank you.

**Comment by nadirhamid** [comment]
Hi so,

'discussion_is_receiver' is in part a request to move 'discussion_content_type'
and 'discussion_content_value' into the 'discussions' table. So:

discussions => 
'discussion_receiver_user_type'
'discussion_receiver_value'

discussion_content =>
'discussion_is_receiver'

**Comment by FWAJL** [comment]
@nadirhamid @bsaiken 

Ok I see now. This is more details than what you previously gave. I will look at it more on Monday.

Regarding upgrade/concern:

1. you can pass on from the Javascript a CSV value for a list of values instead of an array of string. In the controllers, you create the array of string or int values in seconds. No need for an update.
2. Regarding the Dal concern (#842 and #843), we are still waiting for your use case. Please provide it to us.
3. We do have multidimensional array in Session. See this example once logged in with test/test : http://dev.fieldworkmanager.com/debug/view?type=pm or http://dev.fieldworkmanager.com/debug/view?type=project.

Please read this: https://docs.google.com/spreadsheets/d/1HcS6kgDvSaYjffgWfBO7S_Eq3HKsOYfuL-b-0x5Rczw/edit#gid=1372629556
It explains about the Session arrays we use.
I will detail more later if you need.

**Comment by FWAJL** [comment]
@nadirhamid @bsaiken 

Here is my suggestion for the tables structure based on your comments and the progress of understanding the needs:

![discussion thread tables](https://cloud.githubusercontent.com/assets/7524388/7218563/10350e18-e677-11e4-86d9-a712a1f778b6.png)

When a PM selects one FT, it will create a discussion (new row in discussion table, you get a new id) and you will need to insert the two persons interacting (two discussion_person row where the initiator, being the PM, will have the discussion_person_author flag set to 1).

Each discussion person is linked to a user from table user which is then referring to a pm or a technician. It would be the same for any other type of users (service provider, client, etc...). So we don't need the type/value columns as I originally thought.

Then, the PM will compose his message and send it. It will insert a discussion_content row with his discussion_person_id to identify the message's author. This is sent to all the persons via email as you chose (in this example, only one FT but we will support many FTs thanks to the table structures).

Now you can display the messages ordered by discussion_content_id and you simply pull all the discussion_content rows for a discussion.

Does it make sense? I think now it covers our need. Please confirm you can do your work with this. If so, I will make the sql updates.

Thank you.

**Comment by nadirhamid** [comment]
Hi,

Yes this sounds about right, as soon as you have the SQL updated I will
use it in my later draft

**Comment by FWAJL** [comment]
@nadirhamid @bsaiken 

Please see the commit above with all you need to continue.

Also, could you tell us if you understood / have questions about my [comment made earlier](https://github.com/FWAJL/FieldWorkAssistantMVC/issues/803#issuecomment-94037569). We would appreciate an answer from you. Thank you.

**Comment by nadirhamid** [comment]
@bsaiken @FWAJL 

1. I have no need to use the Javascript array anymore, when I do I will once
again look at this comment

2. Recently I have tried to push having received merge conflicts as I have begun doing work on Discussion_person.php and other classes I am unable to put up updates, please let me know of a resolution 
Auto-merging Applications/PMTool/Resources/Enums/ViewVariablesKeys.php
Auto-merging Applications/PMTool/Models/Dao/Discussion_person.php
CONFLICT (add/add): Merge conflict in Applications/PMTool/Models/Dao/Discussion_person.php
Auto-merging Applications/PMTool/Models/Dao/Discussion_content.php
CONFLICT (content): Merge conflict in Applications/PMTool/Models/Dao/Discussion_content.php
Auto-merging Applications/PMTool/Models/Dao/Discussion.php
CONFLICT (content): Merge conflict in Applications/PMTool/Models/Dao/Discussion.php
Auto-merging Applications/PMTool/Models/Dal/DiscussionDal.php
Auto-merging Applications/PMTool/Helpers/DiscussionHelper.php
Removing Applications/PMTool/Controllers/DebugController.php
Removing Applications/PMTool/Config/.appsettings.xml.swp
Automatic merge failed; fix conflicts and then commit the result.

**Comment by bsaiken** [comment]
@nadirhamid @FWAJL 

To commit code, you must follow the instructions at:
https://github.com/FWAJL/FieldWorkAssistantMVC/wiki/3.Use-the-repository-with-Git
They are pretty straightforward.

Always start with a git_pull.  The site is being updated regularly and you will quickly fall behind.  I recommend at least daily, but certainly before you commit any code.

You may only commit code to the feature_thread branch.  So make sure you are on your branch when doing anything.

If you are still having trouble let me know.

It can also helps us if you change the assignee once you make a comment.  A lot is going on.  If you do not change the assignee, Jeremie or I may not notice right away that you want us to look at your comment.

**Comment by FWAJL** [comment]
@nadirhamid @bsaiken 

Like Brian said, you fell behind because as you coded, you didn't do regular pulls. If you are unfamiliar with this issue you are havin, the conflicts exist because I have committed code you should have used to make your dev work by doing a `git pull`.

The 3 files that are problematic are to be merged manually. No other way. You need to check what is yours and what is from the HEAD (or baseline of the file) and do the merge so you don't loose code that is to be kept. You could send us the files if you are having a problem.

Finally, when I merge the dev branch to all feature branches, I had the @mention to the person working on the branch. I did this morning for example for your branch. Have you seen it?

Thank you.

**Comment by nadirhamid** [comment]
@FWAJL  @bsaiken 

Ok. I will look at the changes and pull, it should not be a problem

**Comment by bsaiken** [comment]
@FWAJL @damirius 
Jeremie,
Damir has agreed to take on the Discussion thread Milestone.  Huzzah!  You are currently assigned to this Issue, so I will leave it to you to get him up to speed on where we are.

**Comment by WebDevJL** [comment]
@bsaiken @damirius 

Awesome! 
I will give him the pieces to start working in the morning.
Thank you.

**Comment by FWAJL** [comment]
@damirius 

Please work on the feature_thread branch. The dao classes and SQL updates are what you need and they are all on it. (see commits above).

Here is the SOW: https://docs.google.com/document/d/1cZpNqRRcmy0ASNOrrYDSZypINuZ2wjyNv_EIxOV48F4/edit?usp=sharing (Only the database details have changed and the SOW doesn't reflect this. All the up-to-date details are here: https://github.com/FWAJL/FieldWorkAssistantMVC/issues/803#issuecomment-94248549).

Let me know if you need more info. Thank you.

**Comment by damirius** [comment]
@FWAJL @bsaiken 

Just wanted to let you know that I'm working on this. Interesting issue! I have some fancy ideas, I'll share them when we come to those parts :). First I need to fix some parts of the existing code.

Cheers

**Comment by WebDevJL** [comment]
@damirius @bsaiken 

Let me know if you need database updates as we thought it through and I'm hoping we have all we need. 

Keep up the good work :+1:

**Comment by damirius** [comment]
@FWAJL @bsaiken 

Since you are here, just wanted to make sure that I got this right. When PM clicks on service or FT from the list and then 'promote' it we create new discussion right? So we need separate discussions for each pm->ft/service pair. 
For example PM selects FT named 'John' from the list and starts the discussion, we create new row in the `discussion` table, with current `task_id` and then we create two `discussion_person` entries, one for PM user which is set to 'author' and other one for FT user and of course we link them both to the `discussion` table over `discussion_id` field and set that discussion to the current one.
After that PM selects other FT named 'Gabriel', we do the same thing as with 'John'.
Now we get back and select 'John' again, we need to check pm->ft pair to see if we have existing 'discussion' for them so that we don't create new one.

**Comment by FWAJL** [comment]
@damirius @bsaiken 

What you describe looks right to me. @bsaiken : do you agree?

Technically, I have made the tables for discussion so that we could have many persons instead of 2 only. But right now, we need the most simple solution possible. Evolution will come later.

**Comment by FWAJL** [comment]
@damirius 

BTW, look at #877 for email solution. Let me know if you are familiar with it or if you know another one you prefer.

I am still "designing" the tables right now. Hopefully, I will be ready when you need to send emails.

Thank you.

**Comment by bsaiken** [comment]
@damius @fwajl

That does sounds correct. We wanted to keep it simple - two way conversations only between the PM and someone else. 

We also wanted it to be complete. The PM can talk to anyone involved in the Task. 

People can always phone, text, or email, but this way the information gets archived with the task, and often the conversations are important - especially later.

**Comment by FWAJL** [comment]
@damirius 

I have updated #877 with the issues to complete. Look at #909 for sending email library classes.

**Comment by damirius** [comment]
@FWAJL @bsaiken 

Uh, where to start. I had to rewrite some of the existing stuff we had there. I left communication tab view more or less intact for now. Had to change that startCommWith action we use for setting currentDiscussion and I've added sendMessage action for sending discussion messages (adding discussion_content).

Since we can have multiple discussions for the same task, between pm and every other FT or Service on that task, I had to find a way to see if we already have started discussions in the db. 
So I had to fetch discussion id based on the current task and combination of users.
Here is the query:
`SELECT d.* FROM discussion AS d WHERE d.task_id = :task_id AND :current_user_id IN (SELECT user_id FROM discussion_person AS dp WHERE d.`discussion_id` = dp.discussion_id) AND :recipient_user_id IN (SELECT user_id FROM discussion_person AS dp WHERE d.`discussion_id` = dp.discussion_id)`

I'm not happy about it either :). Jeremie if you have any ideas how to rewrite it please do tell. I'll probably change it later if I come up with something, but it works for now. Basically it tries to find the task discussion for exact pair of users, first one is current user, PM and the other is FT/Service. Also since it is likely that PM will have multiple tasks and maybe same technicians on those tasks too it is filtering by task_id. So discussion between same PM and FT on two different tasks will create two separate discussions.

Send writes the message in `discussion_content` but it isn't sending email for now since we don't have that done, but in any case when we finish that email class implementation it will be easy to send the email in that action.

Now I don't want to send this to function review yet, since I don't know what else needs to be done exactly for this issue and do we want to leave it open till we implement email sending.

I plan to implement history box as the part of the next Issue #804. I'll display there old messages from PM also, we can use some color coding so it will be easier to read.

**Comment by bsaiken** [comment]
@damirius @FWAJL 
Since the email tool is set up as separate Issue (#909), we can continue on this Issue without the email.

Outgoing messages can be typed, stored, and displayed.  Since incoming messages are stored in the database, we could manually populate the database to see a stored incoming message.  That's how we were planning to test it.  

I am still unclear how the incoming message arrives.  Does the user need to be logged into the site, or can they reply to the email (like with github)?  It makes a difference because if this is set up so that messages can only be entered, or "sent" when logged into the site, the contents of email messages need only have a link to log in to the site - in that way the email is simply a notification that there is a message waiting.  I think, for the initial offering of this service, it will only work in places where the Field Technician has the ability to access the internet.  The way things are progressing, both on my site and in the world at large, by the time I am able to offer something that works independent of internet access, internet access will be so widespread that it won't be needed.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

Both outgoing and incoming messages are stored, so we can have 'chat history' box and use it later for 'archived task chat' and so on.

I don't know how will FT dashboard look. If they have some way to look at active tasks that they are assigned to we can add communications tab there too if PM started some discussion with them and then we can show chatbox there too, similar to one which we use for PMs.
They can use it then to read incoming messages and to respond. We can still use email as some kind of notification so they know when they get new message.

**Comment by bsaiken** [comment]
@damirius @FWAJL 
I do not know how the FT dashboard will look either.  That is Issue #850.  It will likely look like whatever you develop - just formatted for a mobile device.

**Comment by FWAJL** [comment]
@bsaiken @damirius 

I'm sorry guys but this is not working. I am using the PM "test" and the project "EC-TX". I have added a FT that has a user row in the DB.
However, I get a "Error reading discussion thread". It is trying to look into the session to find a thread which doesn't exist since I didn't start any.
![screen shot 2015-04-28 at 06 53 27](https://cloud.githubusercontent.com/assets/7524388/7363039/e2886236-ed73-11e4-8066-4b563aa95264.png)

Can you look into it @damirius ? Thank you.

**Comment by bsaiken** [comment]
@FWAJL 
Did you use Damir's database reset script?

**Comment by WebDevJL** [comment]
@damirius @bsaiken 

Where is it Damir's db reset script?
We have a problem if the data is necessary to make this work. It should work with any data.
I will need to take a detailed look at the code to see what is going on.

@damirius : have you pushed everything?

**Comment by damirius** [comment]
@FWAJL @bsaiken 

I sure did pushed everything. If it says "Error reading discussion thread" it's maybe because you didn't start any. You can try to click on the FT and start the thread with him by clicking promote arrow. After that thread should be created.

**Comment by WebDevJL** [comment]
@bsaiken @damirius 

It finally did. I will carry out my code review. Please notify me of any update you make regarding the discussion work as I might do updates as well.

Thank you.

**Comment by FWAJL** [comment]
@damirius 

The code looks ok but I would have just one request for the next code you write. 

1. can you make sure that a controller method doesn't query data from the directly from the database?
2. can you make sure that a controller method doesn't query data from the directly from the session when it needs processing (taking a list or value without processing is fine though)?
3. can you make small methods in the Helpers if you so the code is more reusable? Big methods will create dependencies and it is hard to reuse them.

Otherwise, you are still showing good quality code :+1: Thank you for that.

**Comment by damirius** [comment]
@FWAJL 

I have one problem with discussions and sessions. In most cases in our code when we for example add new tasks or locations as PM, they are directly connected to that user. Only that user can see the tasks/locations, so we use sessions to store read data from the db and we update those session arrays with data when we update or delete that data easily. For discussion threads it's a little different, they are more "dynamic" then our usual stuff. For example I'm PM, I load the discussion and store all messages in the session. FT which I started discussion with responds to some messages and they are added to the db. Since that's totally different user with different session, his session will be updated, but not mine (PMs). Only thing that I can do is to fetch discussion data from the db again to see if there are any new messages.
This problem will occur with other stuff too where two or more users can edit the same things. For example analytes. Admin can add new one, I suppose that will not happen that often as with discussions, but non the less it can. PM already read the analyte data from the db and stored it in the session, so it will not be read again till we log out then log in again, but in the meantime admin added new common analytes. Again we can't know if he added them if we don't check the db.

Other than that I agree that I should add more helper methods for code that I reuse often, I'll see that I move some things if there is need for that.
Thanks for the suggestions/requests, I'll see that I do that.

**Comment by FWAJL** [comment]
@damirius 

In the example of Analytes, we could lock a resource when being edited. That's the first and probably best solution I see. Also we would need to add timestamp in every table to check if DB refresh is necessary.
However, we would need to think about the best implementation and I don't have time for that.

In the example of discussion, let's just refresh the data all the time. However, since we have the messages of the discussion in session, you can query simply on query the last timestamp from the discussion content table for the discussion and if the last timestamp of the messages stored in session is not the same, trigger a refresh (that is 2 helpers methods: 1 to get the last discussion_content timestamp, 1 to refresh the discussion_content data).

I hope this answer your questions.

**Comment by damirius** [comment]
@FWAJL 

I understand, but how much do we get by that? We still query the db every time we need to show messages, just instead of fetching all rows we first check for the latest timestamp then if it is bigger value than timestamp in the session we refresh the data. That's two queries, we can also do it with one where we fetch always all data that have timestamp set to bigger value than the one in session. In either case we query the db and unless we have few some large amount of data in the discussion thread it's the same thing as refreshing all data always.
I don't want to bother you with that anyway, we have bigger things to do. Thanks for the talk.

**Comment by FWAJL** [comment]
> @damirius said:
>  we can also do it with one where we fetch always all data that have timestamp set to bigger value than the one in session.

That's a good solution! You can query the DB for the new data only since you know the last timestamp from the session. Then you update the session data. 
I like that. If you want, do it as it will be more efficient for sure going forward.

**Comment by bsaiken** [comment]
@damirius @FWAJL 
I was reading your discussions and I see that Damir wrote
`Only thing that I can do is to fetch discussion data from the db again to see if there are any new messages.`
Why wouldn't Issue #807 resolve this?

**Comment by damirius** [comment]
@bsaiken @FWAJL 

The thing is that we currently read most of the stuff from database only once and store it in session. After that, when we need that data again we read it from the session. That's OK for data that only one user can edit.
For example we log in as PM and we access technician list. Controller calls some methods to check if we already read technician list before. Since we didn't, controller calls helper which reads the DB and stores all technicians in the session. After that we delete one technician, helper deletes technician in the DB and removes it from the session too. That's great while one user is working on same data.
Discussions are more dynamic, new messages will often be created by 2 users, or even more later on if we choose to add functionality for multiple people conversations.
When we login, for example, as PM again and select some discussion, helper would fetch all messages and store them in the session. After that new messages from PM will be stored in both session and DB, but if FT, which is the part of this discussion, send us some message we will not get it, since we are only reading from sessions and FT can't modify PMs session. So we have to refresh the session too see if there are new messages.
Issue #807 should refresh history chatbox, but refreshing it from the session would not be sufficient we would need to refresh it from the database to see new messages from the other users.
Also any other stuff in the application that can be edited by multiple users will also cause problems.

**Comment by bsaiken** [comment]
@damirius @FWAJL 
We created Issue #807 because we knew this would be a problem.  So why can't the answer to Issue #807 be that the function called by the "button" is to re-initialize the session arrays and refresh the page?  This would be the solution for PMs (only).

For other users, an email notice would be required.  Clients and SPs would not normally be logged in to the site, so they would otherwise never know there was a message.  

We could handle FTs by giving them some kind of alert (e.g. in the header bar).  We would need to have the page auto-refresh every X minutes for them to see the alert.  Then, the FT would have a button that is like the PM button in #807 so they would also re-initialize the session arrays, refresh the page, and see the message.

For that matter, we could add an auto-refresh to the PM UI as a convenient feature.

**Comment by FWAJL** [comment]
@damirius 

Sorry Damir but there is a bug when a PM tried to initiate a new discussion a person (FT or SP). The bug only happens the first time you select to the person and click the arrow. The second time it seems to work fine.

See my screenshot for details:
![screen shot 2015-05-13 at 06 31 32](https://cloud.githubusercontent.com/assets/7524388/7603800/b84cec54-f939-11e4-8a09-c6e3197d4f41.png)

**Comment by bsaiken** [comment]
@damirius @FWAJL 
I was monitoring and did some additional testing.  I was able to re-create the error, but only on the first login attempt with a new task, new FT, and new SP.  Every time after that, there was no error.  I even tried logging out and clearing the cache.  Even that did not re-create the error.


![screen shot 2015-05-12 at 9 44 13 pm](https://cloud.githubusercontent.com/assets/8228624/7603928/3dd7c4c4-f8f0-11e4-92ec-b4d2c8fed845.png)

**Comment by damirius** [comment]
@bsaiken @FWAJL 

This should be fixed now.

**Comment by bsaiken** [comment]
@damirius @FWAJL 
A few loose ends to tie up, but this is working now.  thanks
